common/libraries/util/atecc_utils.c

```diff
@@ -0,0 +1,245 @@
+#include "atecc_utils.h"
+#include "sha2.h"
+#include "flash_api.h"
+#include "controller_level_four.h"
+
+static ATCA_STATUS helper_config_to_sign_internal(ATCADeviceType device_type,
+                                                  struct atca_sign_internal_in_out *param,
+                                                  const uint8_t *config) {
+    const uint8_t *value = NULL;
+    uint16_t slot_locked = 0;
+
+    if (param == NULL || config == NULL || param->temp_key == NULL) {
+        return ATCA_BAD_PARAM;
+    }
+
+    // SlotConfig[TempKeyFlags.keyId]
+    value = &config[20 + param->temp_key->key_id * 2];
+    param->slot_config = (uint16_t) value[0] | ((uint16_t) value[1] << 8);
+
+    // KeyConfig[TempKeyFlags.keyId]
+    value = &config[96 + param->temp_key->key_id * 2];
+    param->key_config = (uint16_t) value[0] | ((uint16_t) value[1] << 8);
+
+    if (device_type == ATECC108A && param->temp_key->key_id < 8) {
+        value = &config[52 + param->temp_key->key_id * 2];
+        param->use_flag = value[0];
+        param->update_count = value[0];
+    } else {
+        param->use_flag = 0x00;
+        param->update_count = 0x00;
+    }
+
+    //SlotLocked:TempKeyFlags.keyId
+    slot_locked = (uint16_t) config[88] | ((uint16_t) config[89] << 8);
+    param->is_slot_locked =
+        (slot_locked & (1 << param->temp_key->key_id)) ? false : true;
+
+    return ATCA_SUCCESS;
+}
+
+void helper_get_gendig_hash(atecc_slot_define_t slot,
+                            uint8_t *data,
+                            uint8_t *digest,
+                            uint8_t *postfix,
+                            atecc_data_t atecc_value) {
```

Comment:
this is wrong. pass by value won't achieve what we want. the data updated by the function would not reflect in the original instance of the struct.

```suggestion
                            atecc_data_t atecc_value) {
```